pkg/endpoint/policy.go

```diff
@@ -333,39 +333,57 @@ func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
 		ExternalDone: make(chan bool),
 	}
 	owner.QueueEndpointBuild(newReq)
-	go func(e *Endpoint, myTurn <-chan bool, finish chan<- bool, externalFinish chan<- bool) {
+	go func(req *Request, e *Endpoint) {
 		buildSuccess := true
+
 		e.Mutex.Lock()
-		e.State = StateRegenerating
+		// If endpoint was marked as disconnected then
+		// it won't be regenerated
+		if e.State != StateDisconnected {
```

Comment:
You are checking `e.State != StateDisconnected` before taking the Mutex. How does this synchronize to `LeaveLocked`? Don't you need to move this into the Mutex?